Add support for 'fixed link' configuration on Kinetics MCUX ENET driver

lmajewski opened the pull request:
Following changes to Kinetics MCUX ENET driver:

- Disable usage of SMI to read status of PHY device.
**Rationale:**
Some devices - like DSA switches (ksz8794) use SPI for configuration and reading its status
(instead of MDIO). The current implementation of ENET driver (in eth_enet.c ) is very tightly coupled with SMI interrupts as they allow moving to the next stage of this driver's state machine. This change decouples driver logic from them when required.

- Add support for 'fixed-link' DTS property
With this patch set the enet driver now supports setting fixed transmission parameters between
PHY (DSA) and ENET. 

Please consider those changes as a preparatory work for DSA support for Zephyr.

Commit:
eth: mcux: Support devices not using SMI for PHY setup

Some ICs - like DSA switches (e.g. ksz8794) - do not use SMI to setup
and configure PHY.

This change introduces a new Kconfig define - CONFIG_ETH_MCUX_NO_PHY_SMI
- to allow replacing SMI communication with SPI or I2C.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
defconfig: ip_k66f: Enable ETH_MCUX_NO_PHY_SMI on ip_k66f board

The ip_k66f board has KSZ8794 connected to K66F's enet, which
requires fixed link operation.


Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
eth: mcux: Add support to NXP's ENET for fixed link setup

This commit adds support for setting fixed configuration, read
from device tree, for ENET ETH interface and PHY.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
dts: Add binding description for ethernet fixed link PHY operation

This patch adds description for 'fixed-link' node in ethernet DTS
node. It supports setting speed and duplex.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
dts: Add 'fixed-link' sub-node to 'eth' node on ip_k66f board

This change adds the 'fixed-link' sub-node of the ip_k66f board's
ethernet node. The fixed link is set to work with 100 Mbps and
full duplex.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

bwasim commented on dts/bindings/ethernet/ethernet,fixed-link.yaml:
```diff
@@ -0,0 +1,26 @@
+      required: false
+      description: The fixed link operates in full duplex mode
+
+    half-duplex:
```
I don't think that "half-duplex" binding is required. The driver should check for "full-duplex" and fallback to half duplex if the property is unavailable.. This seems superfluous.. 

I just looked through the standard device tree bindings for fixed-link and they say the same thing as well (https://android.googlesource.com/kernel/common.git/+/android-trusty-4.4/Documentation/devicetree/bindings/net/fixed-link.txt)

lmajewski replied:
Ok. Thanks for pointing this out - I will follow this recommendation.

bwasim commented on dts/bindings/ethernet/ethernet,fixed-link.yaml:
```diff
@@ -0,0 +1,26 @@
+
+description: Ethernet fixed link configuration for PHY communication
+
+compatible: "ethernet,fixed-link"
```
If you look at the standard device tree bindings for "fixed-link", there is no compatible property.. Whenever the Ethernet device is enumerated, the driver is supposed to search for the "fixed-link" node and assume that the driver uses fixed link if found.. See https://elixir.bootlin.com/linux/v4.5/source/drivers/of/of_mdio.c#L376 for details.. 

I'm not sure what exactly is the policy about introducing "zephyr specific" bindings, but this seems avoidable..

lmajewski replied:
On the beginning I wanted to follow the same pattern - i.e. to have the 'fixed-link' parsed as a subnode.
However, I've encountered the following issue:
- In Zephyr I use DT_CHILD(DT_ALIAS(eth), fixed_link) to get the information if 'fixed-link' property is available.
- However, there is the issue - without proper ethernet,fixed-link.yaml and compatible node
the parsed DTS (in zephyr/include/generated/devicetree_unfixed.h) doesn't have required
defines (for example #define DT_N_S_soc_S_ethernet_400c0004_S_fixed_link*).
Without them the DT_CHILD(DT_ALIAS(eth), fixed_link) macro resolve to 0.

Is there any way to avoid having "compatible" property and have required defines for DT_CHILD* macros?

jukkar commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -238,6 +250,24 @@ static int eth_mcux_device_pm_control(struct device *dev, u32_t command,
+static void eth_mcux_fixed_phy_duplex_and_speed(phy_duplex_t *p_phy_duplex,
+						phy_speed_t *p_phy_speed)
+{
+
```
Remove extra empty line

lmajewski replied:
Ok

jukkar commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -471,8 +504,12 @@ static void eth_mcux_phy_event(struct eth_context *context)
 	case eth_mcux_phy_state_read_status:
 		/* PHY Basic status is available. */
+#if defined(CONFIG_ETH_MCUX_NO_PHY_SMI) && ETH_MCUX_FIXED_LINK
+		link_up = 1;
```
As the variable is pool, do `link_up = true;` instead

lmajewski replied:
Ok.

jukkar reviewed (changes requested):
Please fix also checkpatch warnings.

lmajewski commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -391,6 +405,17 @@ static void eth_mcux_phy_event(struct eth_context *context)
+		 * otherwise reschedule the eth_mcux_phy_event and check after
+		 * 1ms
+		 */
+		if (context->iface)
```
Ok.

galak commented:
I need to look at the DT stuff, but think it can be handled possibly as a child.  you can see something like dts/bindings/mtd/partition.yaml

Commit:
dts: Support 'fixed-link' property in nxp,kinetics-ethernet.yaml

This commit enables support for parsing 'fixed-link' node when it
is added to node described in 'ethernet,fixed-link.yaml'.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

pfalcon commented:
@lmajewski: I usually all for having well-split, dine-detail commits, but sometimes, you can split too much. Your commit "eth: mcux: Support devices not using SMI for PHY setup" says in description "This change introduces a new Kconfig define - CONFIG_ETH_MCUX_NO_PHY_SMI", but it actually doesn't, and thus references an unknown CONFIG_* option. Bottom line: please squash "eth: Kconfig: Add new ETH_MCUX_NO_PHY_SMI Kconfig option" into that commit. (The rest seems to be ok.)

pfalcon commented:
@lmajewski: Also, while doing the above, please consider whether it's worth to name the option CONFIG_ETH_MCUX_**NO**_PHY_SMI, maybe it makes sense to cut the negation, name it CONFIG_ETH_MCUX_PHY_SMI, have it "y" be default, and just disable for platforms where needed. Well, might worth asking @MaureenHelm whether she thinks that's anyhow important.

pfalcon commented:
@jukkar: It seems that your previous comments were addressed, can you have a look and maybe dismiss your -1 to motivate other reviewers to look further into this?

lmajewski commented:
> @lmajewski: I usually all for having well-split, dine-detail commits, but sometimes, you can split too much. Your commit "eth: mcux: Support devices not using SMI for PHY setup" says in description "This change introduces a new Kconfig define - CONFIG_ETH_MCUX_NO_PHY_SMI", but it actually doesn't, and thus references an unknown CONFIG_* option. Bottom line: please squash "eth: Kconfig: Add new ETH_MCUX_NO_PHY_SMI Kconfig option" into that commit. (The rest seems to be ok.)

I will squash those commits - thanks for pointing this out

> @lmajewski: Also, while doing the above, please consider whether it's worth to name the option CONFIG_ETH_MCUX_**NO**_PHY_SMI, maybe it makes sense to cut the negation, name it CONFIG_ETH_MCUX_PHY_SMI, have it "y" be default, and just disable for platforms where needed. Well, might worth asking @MaureenHelm whether she thinks that's anyhow important.

As the eth_mcux.c driver uses PHY_SMI for all NXP SoCs excluding only DSA case - I think that it is more readable to use straight logic and put all the necessary code under 
`#ifdef CONFIG_ETH_MCUX_NO_PHY_SMI`

Otherwise, we would have negation in the `#ifdef` - `#ifndef` which IMHO is far more difficult to read correctly. Additionally, with current approach we can **only** enable this option when we do need it - in the DSA case.

**To sum up:**
I would prefer to keep the name CONFIG_ETH_MCUX_**NO**_PHY_SMI (and current approach) as it shows that this code doesn't use
SMI interrupts for eth_mcux operation.

pfalcon commented:
> Otherwise, we would have negation in the #ifdef - #ifndef which IMHO is far more difficult to read correctly.

I would disagree with that, it's easier to deal with standard language syntax of `#ifdef`/`#ifndef` which applies to any situation, than to adhocly named options which embed some kind of negation somewhere in their names.

That said, as long as that option is considered, and you have arguments to prefer the current situation, it sounds good to me.

lmajewski commented:
> > Otherwise, we would have negation in the #ifdef - #ifndef which IMHO is far more difficult to read correctly.
> 
> I would disagree with that, it's easier to deal with standard language syntax of `#ifdef`/`#ifndef` which applies to any situation, than to adhocly named options which embed some kind of negation somewhere in their names.
> 
> That said, as long as that option is considered, and you have arguments to prefer the current situation, it sounds good to me.

Please correct me if I've misunderstood something - you are OK to keep things as they are proposed in this patch set (as I've pointed out arguments for such situation). 

However, you disagree with the generic approach to have CONFIG_* names with negation in them and you prefer to use `#ifndef` instead when applicable?

pfalcon commented:
> Please correct me if I've misunderstood something - you are OK to keep things as they are proposed in this patch set (as I've pointed out arguments for such situation).

Yes.

> However, you disagree with the generic approach to have CONFIG_* names with negation in them and you prefer to use #ifndef instead when applicable?

Yes, I just share my personal opinion on that matter. Maybe you'll find some merit in it and consider it next time (maybe not). Likewise, I'm listening to arguments of others. All that should help make the project better/more consistent/easier to use for everyone.

Bottom line: please make the commit squash discussed above. Then, from my side, I'd just like to test it on frdm_k64f (keeping fingers crossed that I'll have time for that this week).

agansari reviewed (commented):
Code looks okay, waiting for @pfalcon to test it to approve.

My comment related to Kconfig: I find the double negation sometimes harder to follow
`#ifndef CONFIG_ETH_MCUX_NO_PHY_SMI`
In logic a double negation is a truth statement, but it's not as clear and straight forward as a direct truth claim when following code.

pfalcon reviewed (approved):
Sorry for the delay here, tested on frdm_k64f with my usual tests - dumb_http_server, total ~5000 requests, and big_http_download, 3 iterations, total 20MB downloaded.

Ideally, I wish this was reviewed/approved by @agansari and/ot @MaureenHelm, but as this waited in queue for so long, let me cast +1 from me now.

MaureenHelm reviewed (approved):
I agree the double negative is confusing and I generally prefer to see positive logic in Kconfigs. But given that 3 networking folks have +1-ed this, I'm ok to keep it as it is.
